### ISSUE TITLE
CTE in update_all

### ISSUE THREAD

**Issue by mbuhot** [elixir-ecto/ecto#3052]
### Precheck

* Do not use the issues tracker for help or support requests (try Elixir Forum, Stack Overflow, IRC or mailing lists, etc).
* For proposing a new feature, please start a discussion on [elixir-ecto](https://groups.google.com/forum/#!forum/elixir-ecto).
* For bugs, do a quick search and make sure the bug has not yet been reported.
* Finally, be nice and have fun!

Apologies if this is more feature request than bug report. I tried to access the google groups link above, but got this error: 

```
This group either doesn't exist, or you don't have permission to access it. If you're sure this group exists, contact the Owner of the group and ask them to give you access.
```

### Environment

* Elixir version (elixir -v):

```
Erlang/OTP 21 [erts-10.2.3] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [hipe] [dtrace]

Elixir 1.8.1 (compiled with Erlang/OTP 21)
```

* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): postgres:9.6
* Ecto version (mix deps): ecto 3.2.0-dev at 8fa9df5 ecto_sql 3.2.0-dev  at 542fe4d
* Database adapter and version (mix deps): postgrex 0.14.3
* Operating system: MacOS 10.14.5

### Current behavior

I'd like to express the following update statement using Ecto's recently added CTE support:

```sql
WITH available_jobs AS (
    SELECT id 
    FROM jobs
    WHERE jobs.state = 'AVAILABLE'
    ORDER BY jobs.id ASC
    LIMIT 10
    FOR UPDATE SKIP LOCKED
)
UPDATE jobs
SET state = 'RESERVED'
FROM available_jobs
WHERE jobs.id = available_jobs.id
RETURNING jobs.*
```

The equivalent ecto code:

```elixir
available_jobs = 
  from(
    job in Jobs,
    where: job.state == "AVAILABLE",
    order_by: [asc: job.id],
    limit: 10,
    lock: "FOR UPDATE SKIP LOCKED",
    select: %{id: job.id}
  )

Job
|> with_cte("available_jobs", as: ^available_jobs)
|> join(:inner, [job], a in "available_jobs", on: job.id == a.id) 
|> select([job, _available], job)
|> Repo.update_all(set: [state: "RESERVED"])
```

Fails with an error: 

```
** (Ecto.QueryError) `update_all` allows only `where` and `join` expressions. You can exclude unwanted expressions from a query by using Ecto.Query.exclude/2. Error found in query:
     
     from j0 in Jobs,
       join: a1 in "available_jobs",
       on: j0.id == a1.id,
       update: [set: [state: ^"RESERVED"]],
       select: j0
```

### Expected behavior

Ideally an `update_all` with a query that joins on a CTE should behave the same as with a query that joins on a subquery, but with the expectation that the CTE is materialized only once by the database.

So the code above should be handled by ecto similarly to:

```elixir
Job
|> join(:inner, [job], a in subquery(available_jobs), job.id == a.id)
|> select([job, _a], job)
|> Repo.update_all(set: [state: "RESERVED"])
```

Unfortunately the above is susceptible to the postgres query planner re-evaluating the subquery multiple times, producing more results than the subquery limit intended, [eg](https://github.com/feikesteenbergen/demos/blob/master/bugs/update_from_correlated.adoc).

I'm happy to try to contribute support for this given some direction on why `update_all` queries are restricted to only allowing `where` and `join` expressions.

Thanks

**Comment by josevalim** [comment]
There is no reason why we support only joins and where in updates besides sanity. We want to make sure new behaviours work intentionally and not by accident. So we would definitely appreciate two prs:

1. One that changes ecto to allow with_ctes in updates/deletes
2. One that changes ecto_sql to consider with_ctes with generating update_all/delete_all queries and add new tests in `test/ecto/adapters/*_test.exs`

**Comment by mbuhot** [comment]
Thanks @josevalim. I've created a couple of PRs. Could you provide some guidance on the changes to `Ecto.Query.Planner` to properly manage caching and parameter handling for update_all queries with CTEs?

**Comment by mbuhot** [comment]
Resolved by #3053 and https://github.com/elixir-ecto/ecto_sql/pull/125

Thanks for the assistance @josevalim!
